python_modules/dagster/dagster/_core/definitions/asset_graph.py

```diff
@@ -262,18 +199,30 @@ def get_asset_partitions_in_range(
             for partition_key in partition_keys_in_range
         ]
 
-    def get_partition_mapping(
-        self, asset_key: AssetKey, in_asset_key: AssetKey
-    ) -> PartitionMapping:
-        partition_mappings = self._partition_mappings_by_key.get(asset_key) or {}
-        return infer_partition_mapping(
-            partition_mappings.get(in_asset_key),
-            self.get_partitions_def(asset_key),
-            self.get_partitions_def(in_asset_key),
-        )
-
     def is_partitioned(self, asset_key: AssetKey) -> bool:
-        return self.get_partitions_def(asset_key) is not None
+        # Temporarily performing an existence check here for backcompat. Callsites need to be
+        # changed to verify this first.
+        return asset_key in self.all_asset_keys and self.get_partitions_def(asset_key) is not None
+
+    @abstractmethod
+    def get_freshness_policy(self, asset_key: AssetKey) -> Optional[FreshnessPolicy]:
+        ...
+
+    @abstractmethod
+    def get_auto_materialize_policy(self, asset_key: AssetKey) -> Optional[AutoMaterializePolicy]:
+        ...
+
+    @abstractmethod
+    def get_auto_observe_interval_minutes(self, asset_key: AssetKey) -> Optional[float]:
+        ...
+
+    @abstractmethod
+    def get_backfill_policy(self, asset_key: AssetKey) -> Optional[BackfillPolicy]:
+        ...
+
+    @abstractmethod
+    def get_code_version(self, asset_key: AssetKey) -> Optional[str]:
+        ...
```

Comment:
Retyping this one.

So this is an opportunity to introduce a new class that represents an asset's metadata after it is construct and bound in a repo. Meaning it is different than an `AssetSpec`. Call it `AssetInfo` or `AssetNode`.

```python
class AssetNode:
   deps: List[AssetNode]
   freshness_policy: FreshnessPolicy
   code_version: Optional[str]
   # and so on and so forth
```

Now instead of 

```asset_graph.get_code_version(asset_key)```

You would have 

```asset_graph.get_asset_node(asset_key).code_version```

There are a number of advantages here

1) It reduces the surface area of the AssetGraph API considerably. `get_code_version`. `get_backfill_policy` etc are all gone. As are `is_executable`. 
2) If we add properties to `AssetNode` we do not need to alter `AssetGraph`. I.e. it actually composes

Reply:
I agree with this direction but would like to proceed incrementally, since there is already significant work stacked on this PR that:

(1) eliminates reliance on the implicit assumption that undefined asset keys are valid source assets: https://github.com/dagster-io/dagster/pull/19973
(2) purges source asset from `AssetGraph`: https://github.com/dagster-io/dagster/pull/19901

If this PR and the above two land then we are in a good state to move to using node objects.

Also reposting my comment from DMs on node objects here for visibility:

Here is what I’m thinking:

```### Abstract classes

class AssetNode(ABC):
    ...

T_AssetNode = TypeVar("T_AssetNode", bound=AssetNode)

class AssetGraph(ABC, Generic[T_AssetNode]):
    ...

### Code-location scoped classes (currently `InternalAssetGraph`)

class LocalAssetNode(AssetNode):  # or SingleLocationAssetNode
    ...

class LocalAssetGraph(AssetGraph[LocalAssetNode]):  # or SingleLocationAssetGraph
    ...

### Multi-code-location scoped classes (currently `ExternalAssetGraph`)

class GlobalAssetNode(AssetNode):  # or MultiLocationAssetNode
    ...

class GlobalAssetGraph(AssetGraph[GlobalAssetNode]):  # or MultiLocationAssetGraph
    ...```

Note that even though we currently have `ExternalAssetNode`, this is single-location-scoped. This is a confusing mismatch with `ExternalAssetGraph`, which is multi-location scoped.

Reply:
Ok. Per offline discussion we can move forward here and defer the `AssetNode` addition, but with the understanding that we do that sooner rather than later.